### PR TITLE
[FLINK-26583][runtime] Make the Flink cluster fail if the job ID is listed as globally-terminated and cleaned

### PR DESCRIPTION
**Pull request by XComp**
## What is the purpose of the change

The Application and Job Mode rely on the `JobResultStore` for recovery. Nothing will be recovered if the job is in dirty state but globally terminated. That will just trigger the cleanup. But we didn't cover the case where there's a clean entry in the `JobResultStore`. This lead to the Application just terminating successfully without any indication to the user that the job didn't run.

~This PR changes the behavior in a way that a `DuplicateJobSubmissionException` with a meaningful error message is thrown instead.~

This PR adds a warning in cases where an job with an ID is submitted that refers to an already globally-terminated and cleaned job.

## Brief change log

* Adds check and log message in `Dispatcher.submitJob`

## Verifying this change

* Adds test in `DispatcherTest` to cover the log message path

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java [hunk 1]
```diff
@@ -77,7 +79,15 @@ public JobDispatcherLeaderProcessFactory createFactory(
         }
 
         final JobResultStore jobResultStore = jobPersistenceComponentFactory.createJobResultStore();
-        final Collection<JobResult> recoveredDirtyJobResults = getDirtyJobResults(jobResultStore);
+
+        if (accessJobResultStore(
+                () -> jobResultStore.hasCleanJobResultEntry(jobGraph.getJobID()))) {
+            throw new FlinkRuntimeException(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java:85]: Seems questionable to throw a RuntimeException here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java [hunk 2]
```diff
@@ -77,7 +79,15 @@ public JobDispatcherLeaderProcessFactory createFactory(
...
+                    DuplicateJobSubmissionException.ofGloballyTerminatedAndCleaned(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java:86]: Why do we check this in 2 places (here and the ApplicationDispatcherBootstrap)?

It seems like we keep spreading these checks around; Dispatcher#submitJob is alkso checking for duplicate jobs...
Not sure whether this is maintainable.

**Review by zentol — COMMENTED** [review]
I don't quite understand get why we should fail.

Like...isn't it _good_ to just skip the job execution? Sounds like idempotent job submissions to me, don't we need that if we want to support application mode with multiple jobs?

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java [hunk 3]
```diff
@@ -1361,8 +1364,15 @@ public void onFatalError(Throwable exception) {
...
 
+        private final Consumer<Throwable> customFatalErrorHandler;
+
+        public ShutDownFatalErrorHandler(Consumer<Throwable> customFatalErrorHandler) {
+            this.customFatalErrorHandler = customFatalErrorHandler;
+        }
+
         @Override
         public void onFatalError(Throwable exception) {
+            customFatalErrorHandler.accept(exception);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:1375]: since isn't really an error handler, for the simple reason that the ShutDownFatalErrorHandler is still called and will still kill the mini cluster. rephrase it to "fatalErrorListener".

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java [hunk 4]
```diff
@@ -165,6 +168,69 @@ public void testDispatcherRecoversAfterLosingAndRegainingLeadership() throws Exc
...
+                        TestingJobResultStore.createSuccessfulJobResult(
+                                ApplicationDispatcherBootstrap.ZERO_JOB_ID)));
+
+        final CompletableFuture<Throwable> fatalError = new CompletableFuture<>();
+        testJobResultStoreEntryLeadingToDuplicateJobSubmissionException(
+                jobResultStore,
+                ErrorHandlingSubmissionJob.createPackagedProgram(),
+                fatalError::complete);
+
+        FlinkAssertions.assertThatChainOfCauses(ErrorHandlingSubmissionJob.getSubmissionException())
```
Comment (zentol) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java:185]: The fact that we have to rely on this singleton is a cause for concern. There should be a better way to retrieve the job result when using a mini cluster with application mode.

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java [hunk 5]
```diff
@@ -165,6 +168,69 @@ public void testDispatcherRecoversAfterLosingAndRegainingLeadership() throws Exc
...
+                .isFalse();
+        assertThat(
+                        jobResultStore.hasCleanJobResultEntry(
+                                ApplicationDispatcherBootstrap.ZERO_JOB_ID))
+                .isTrue();
+    }
+
+    @Test
+    public void testWithCleanJobResultInApplicationMode() throws Exception {
+        final JobID jobId = ApplicationDispatcherBootstrap.ZERO_JOB_ID;
```
Comment (zentol) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java:204]: Does it matter that its ZERO_JOB_ID? If not, why not just use some JobID?

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java [hunk 6]
```diff
@@ -165,6 +168,69 @@ public void testDispatcherRecoversAfterLosingAndRegainingLeadership() throws Exc
...
+
+        final AtomicReference<Throwable> fatalErrorRef = new AtomicReference<>();
+        testJobResultStoreEntryLeadingToDuplicateJobSubmissionException(
+                jobResultStore, FailingJob.getProgram(), fatalErrorRef::set);
+
+        final DuplicateJobSubmissionException expectedException =
+                DuplicateJobSubmissionException.ofGloballyTerminatedAndCleaned(jobId);
+        assertThat(fatalErrorRef.get())
+                .isNotNull()
+                .as("A fatal error should have caused the Flink cluster to shut down.")
```
Comment (zentol) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java:220]: if this weren't the case then testJobResultStoreEntryLeadingToDuplicateJobSubmissionException wouldn't return in the first place, no?

File: flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java [hunk 7]
```diff
@@ -265,6 +265,35 @@ public void testApplicationFailsAsSoonAsOneJobFails() throws Throwable {
...
+
+        final ApplicationDispatcherBootstrap bootstrap =
+                new ApplicationDispatcherBootstrap(
+                        getProgram(1),
+                        Collections.emptyList(),
+                        actualJobId -> true,
+                        configuration,
+                        // passing null should not lead to any problem because we want to fail
+                        // before doing any call to the dispatcher
+                        null,
```
Comment (zentol) [inline flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java:282]: eh. Why should this test fail if I eagerly retrieve the address from the dispatcher gateway?

**Comment by XComp** [comment]
The rational behind this change is to make the user aware of the fact that there's still a `JobResultEntry` laying around. Flink itself is behaving as expected. Alternatively, we could add a warning. I was just afraid that the user might not notice it and, therefore, went for the exception approach, instead.

File: flink-runtime/src/main/java/org/apache/flink/runtime/client/DuplicateJobSubmissionException.java [hunk 8]
```diff
@@ -35,10 +35,24 @@ public static DuplicateJobSubmissionException ofGloballyTerminated(JobID jobId)
         return new DuplicateJobSubmissionException(jobId, true);
     }
 
+    public static DuplicateJobSubmissionException ofGloballyTerminatedAndCleaned(JobID jobId) {
+        return new DuplicateJobSubmissionException(
+                jobId,
+                true,
+                "There's already a job with the given JobID '"
+                        + jobId
+                        + "' listed as globally-terminated and cleaned. No further actions required.");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/client/DuplicateJobSubmissionException.java:44]: I'm not sure if this is sufficient.

It says "No further actions required", but I'm not sure whether this refers to Flink ("I don't need to run anything") or the user ("I can ignore this error").
Then, let's say a user really wants to run the job anyway; how do they achieve that? (This is particularly important because ZERO_JOB_ID exists)

**Comment by zentol** [comment]
> The rational behind this change is to make the user aware of the fact that there's still a JobResultEntry laying around. Flink itself is behaving as expected. Alternatively, we could add a warning. I was just afraid that the user might not notice it and, therefore, went for the exception approach, instead.

Will in practice, if say you use Kubernetes, this not result in the job being re-submitted again and again because the cluster fails until some failure rate policy is triggered? I'm not sure if this is the better alternative.
In particular because this can happen without the user doing something wrong; say the JM crashes after having cleaned the job result. In that situation we very much want Flink to just shut down.

**Comment by XComp** [comment]
We agreed on going for the log-warning approach because we don't want to cause a failover of the Flink cluster in such a case where the job is already marked as clean. Flink behaves as expected. The user should only be informed about it through logs.

Additionally, we can improve the change by moving the check into `Dispatcher.submitJob` because that's called by all the different execution modes. I reverted the initial commits and pushed the changes to the remote branch. PTAL

**Comment by XComp** [comment]
I added a test case to make the log message reproducible in a unit test.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 9]
```diff
@@ -428,6 +428,12 @@ private void stopDispatcherServices() throws Exception {
 
         try {
             if (isDuplicateJob(jobGraph.getJobID())) {
+                if (isGloballyTerminatedAndCleanedUp(jobGraph.getJobID())) {
+                    log.warn(
+                            "A job having the ID {} is already in globally-terminal state and marked as clean in the JobResultStore. No execution is going to happen.",
+                            jobGraph.getJobID());
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:434]:
```suggestion
                    log.warn(
                            "Ignoring JobGraph submission '{}' ({}) because the job already reached a {} state in a previous execution.",
                            jobGraph.getName(), jobGraph.getJobID(), <insert job state>);
```

**Review by zentol — APPROVED** [review]

**Comment by XComp** [comment]
Unstable e2e profiles are independent of the changes of this PR.
